src/utils/simulation.js

```diff
@@ -0,0 +1,193 @@
+   * call the methods: time_step, kill_or_vaccinate_, infect_newly_infected.
+   */
+  run() {
+    let time_steps = 0;
+    //console.log("run should", this.simulation_should_continue())
+    while (this.simulation_should_continue()) {
+      this.time_step();
+      time_steps += 1;
+      this.kill_or_vaccinate();
+      this.infect_newly_infected();
+    }
+    //console.log("run should2", this.simulation_should_continue())
+    console.log("still sick", this.population.filter((person) => person.infected))
+    console.log(`The simulation has ended after ${time_steps} turns.`);
+  }
+
+  /**
+   * It provides us randomly selected people we can interact with. 
+   * @param {array} alive - People who are alive and not vaccinated.
+   * @param {number} count - The amount of people we want from the alive array (interactions needed).
+   * @returns {array} sample - Ids of people who can get sick.
+   */
+  interaction_sample (alive, count) {
+    let sample = [];
+    let counter = 0;
+
+    while (counter < count) {
+      const random_num = Math.floor(Math.random() * alive.length);
+      const alive_person= alive[random_num];
+      // checks sample array to see if alive person is not in array
+      if (!sample.some(person =>  person === alive_person)) {
+        sample.push(alive_person._id);
+        counter += 1
+      }
+    }
+    return sample;
+  }
+
+  /**
+   * Each infected person is passed to the getter method getAlivePeople to generate a list of people that can still get infected.
+   * If healthy people is less than the total interactions, the infected person interacts will all healthy people. 
+   * If healthy people is more than the total interactions, we generate an array with the amount needed then infected
+   * person interacts with each one. 
+   */
+  time_step() {
+    this.current_infected.forEach((infected) => {
+      // list of people that can get infected
+      let healthy_people = this.getAlivePeople(infected);
+      // if less than total interactions needed => interact with all
+      if (healthy_people.length <= this.total_interactions) {
+        for (let i = 0; i < healthy_people.length; i++) {
+          const healthy_person = healthy_people[i];
+          // check that person was not previously interact and infected during same timestep
+          if (this.newly_infected.indexOf(healthy_person._id) === -1) {
+            this.interaction(healthy_person._id);
+          } 
+        }
+      } else {
+        // get interaction sample with method interaction_sample
+        const interactions = this.interaction_sample(healthy_people, this.total_interactions);
+        interactions.forEach((rand_healthy_person_id) => {
+          // if random person is not already in newly_infected => interaction
+          if (this.newly_infected.indexOf(rand_healthy_person_id) === -1) {
+            this.interaction(rand_healthy_person_id);
+          }  
+        })
+      }
```

Comment:
```suggestion
              for (let i = 0; i < healthy_people.length; i++) {
          const healthy_person = healthy_people[i];
          // check that person was not previously interact and infected during same timestep
          if (this.newly_infected.indexOf(healthy_person._id) === -1) {
            this.interaction(healthy_person._id);
          } 
        }
      } else {
        // get interaction sample with method interaction_sample
        const interactions = this.interaction_sample(healthy_people, this.total_interactions);
        interactions.forEach((rand_healthy_person_id) => {
          // if random person is not already in newly_infected => interaction
          if (this.newly_infected.indexOf(rand_healthy_person_id) === -1) {
            this.interaction(rand_healthy_person_id);
          }  
        })
      }
```

redundant loops. use the if to setup which array you want to iterate through since both loops do the exact same operation

Reply:
Fixed